internal/provider_fm/resource_id.go

```diff
@@ -0,0 +1,219 @@
+used as unique identifiers for other resources.
+
+This resource *does* use a cryptographic random number generator in order
+to minimize the chance of collisions, making the results of this resource
+when a 16-byte identifier is requested of equivalent uniqueness to a
+type-4 UUID.
+
+This resource can be used in conjunction with resources that have
+the ` + "`create_before_destroy`" + ` lifecycle flag set to avoid conflicts with
+unique names during the brief period where both the old and new resources
+exist concurrently.
+`,
+		Attributes: map[string]tfsdk.Attribute{
+			"keepers": {
+				Description: "Arbitrary map of values that, when changed, will trigger recreation of " +
+					"resource. See [the main provider documentation](../index.html) for more information.",
+				Type: types.MapType{
+					ElemType: types.StringType,
+				},
+				Optional:      true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{tfsdk.RequiresReplace()},
+			},
+			"byte_length": {
+				Description: "The number of random bytes to produce. The minimum value is 1, which produces " +
+					"eight bits of randomness.",
+				Type:          types.Int64Type,
+				Required:      true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{tfsdk.RequiresReplace()},
+			},
+			"prefix": {
+				Description: "Arbitrary string to prefix the output value with. This string is supplied as-is, " +
+					"meaning it is not guaranteed to be URL-safe or base64 encoded.",
+				Type:          types.StringType,
+				Optional:      true,
+				PlanModifiers: []tfsdk.AttributePlanModifier{tfsdk.RequiresReplace()},
+			},
+			"b64_url": {
+				Description: "The generated id presented in base64, using the URL-friendly character set: " +
+					"case-sensitive letters, digits and the characters `_` and `-`.",
+				Type:     types.StringType,
+				Computed: true,
+			},
+			"b64_std": {
+				Description: "The generated id presented in base64 without additional transformations.",
+				Type:        types.StringType,
+				Computed:    true,
+			},
+			"hex": {
+				Description: "The generated id presented in padded hexadecimal digits. This result will " +
+					"always be twice as long as the requested byte length.",
+				Type:     types.StringType,
+				Computed: true,
+			},
+			"dec": {
+				Description: "The generated id presented in non-padded decimal digits.",
+				Type:        types.StringType,
+				Computed:    true,
+			},
+			"id": {
+				Description: "The generated id presented in base64 without additional transformations or prefix.",
+				Type:        types.StringType,
+				Computed:    true,
+			},
+		},
+	}, nil
+}
+
+func (r resourceIDType) NewResource(_ context.Context, p tfsdk.Provider) (tfsdk.Resource, diag.Diagnostics) {
+	return resourceID{
+		p: *(p.(*provider)),
+	}, nil
+}
+
+type resourceID struct {
+	p provider
+}
+
+func (r resourceID) Create(ctx context.Context, req tfsdk.CreateResourceRequest, resp *tfsdk.CreateResourceResponse) {
+	if !r.p.configured {
```

Comment:
I know you saw this in the scaffolding too, but I wonder why do we need this pattern: don't we have guarantees that a callstack, reaching this method, definitely has `p.configured` true?

Reply:
That's a good question. I'm not sure. Perhaps @bflad can steer us in the right direction?

Reply:
While the `ConfigureProvider` RPC should always occur before `ApplyResourceChange` RPCs, it is possible that the provider configuration contains unknown values during the `ConfigureProvider` call. For example, a configuration may create an AWS RDS instance resource, then immediately reference attribute values of that resource in the provider configuration for say, a MySQL or Postgres provider that then has its own resources. Terraform CLI will typically perform provider configurations early in the graph for various reasons, not necessarily slotting in the provider configuration after the dependency resource creation.

Most providers have opted for the simpler setup of complaining during provider configuration if there are unknown values, rather than attempting to defer provider configuration logic to the last possible moment. There is limitations in that deferral situation, since the provider configuration is not present in other RPCs. I'm guessing most, if not all, of the utility providers can do the same and skip the configuration check during CRUD, since they typically don't have provider configuration at all anyways. Of course, this can be changed in the future if necessary.

Reference: https://github.com/hashicorp/terraform/issues/4149
Reference: https://github.com/hashicorp/terraform/issues/30937